lib/block_orthogonalize.cu

```diff
@@ -1,188 +1,211 @@
 #include <color_spinor_field.h>
-#include <tune_quda.h>
 #include <uint_to_char.h>
 #include <vector>
 #include <assert.h>
 #include <utility>
 
-#include <launch_kernel.cuh>
-#include <jitify_helper.cuh>
 #include <kernels/block_orthogonalize.cuh>
+#include <tunable_block_reduction.h>
 
 namespace quda {
 
+  struct OrthoAggregates {
+    // List of block sizes we wish to instantiate.  The required block
+    // size is equal to number of fine points per aggregate, rounded
+    // up to a whole power of two.  So for example, 2x2x2x2 and
+    // 3x3x3x1 aggregation would both use the same block size 32
+#ifndef QUDA_FAST_COMPILE_REDUCE
+    static constexpr std::array<unsigned int, 6> block = {32, 64, 128, 256, 512, 1024};
+#else
+    static constexpr std::array<unsigned int, 1> block = {1024};
+#endif
+
+    /**
+       @brief Return the first power of two block that is larger than the required size
+    */
+    static unsigned int block_mapper(unsigned int raw_block)
+    {
+      for (auto block_ : block) if (raw_block <= block_) return block_;
+      errorQuda("Invalid raw block size %d\n", raw_block);
+      return 0;
+    }
+  };
+
+#ifndef QUDA_FAST_COMPILE_REDUCE
```

Comment:
What's the purpose of having these declarations below the definitions above?

Reply:
These are needed pre-C++17 since static constexpr variables are not inlined.  If you don't have these definitions then you'll get link errors when building in debug mode (same reason we have the definitions in instantiate.cpp).  We can delete these one we go C++17.